corsair-psu.c

```diff
@@ -575,6 +587,13 @@ static int corsairpsu_probe(struct hid_device *hdev, const struct hid_device_id
 		dev_err(&hdev->dev, "unable to query firmware (%d)\n", ret);
 		goto fail_and_stop;
 	}
+	/* TODO: decide how to deal with failure to read these values */
```

Comment:
How do you think failures here should be handled? We don't want to cache a zero, that's misleading to the user. A printk also didn't feel appropriate (at least by itself). I also considered caching the values inside of corsairpsu_get_value so that a read error could be returned to userspace appropriately, and userspace could decide if they wanted to retry the read. I felt that a retry loop (especially an infinite one) was not appropriate for a module's *_probe function.

Reply:
First, never use printk, there is pr_info, pr_warn etc now, if you have a device to work on, go for the dev_ functions (dev_info, dev_err etc). I would query all the constants be part of the init, if it fails, I consider that a faulty/non-working hardware. The reason for this is, that micro-controller is not 100% stable (it depends on the hw revision and the firmware). I managed to even lock-up the micro-controller and had to poweroff the psu completely to get it working again. There is also the issue, that not all commands/values are supported by all devices in the RMi/HXi series (it also depends on hw revision and firmware version). There is a reason I left this out. Actually the subsystem maintainer warned me about this, that there will be people who want to add more and more stuff (also unstable/untested stuff) to the driver. ;-) So it is fine, if you want to keep this in a fork, but getting it mainlined will be difficult.

Reply:
Actually, if you want to extract/get more info from the psu or even configure it, do it in userspace. I will provide a small testing tool which I use to digging into the AXi series and which has combined support for all 3 series.